src/Talis/Persona/Client/Tokens.php

```diff
@@ -56,51 +60,66 @@ public function validateToken($params = [])
-            // if the certificate is invalid. Calling openssl_pkey_get_public
-            // first ensures that the certificate is valid before progressing.
-
-            if ($cert = openssl_pkey_get_public($rawCert)) {
-                $decoded = (array)JWT::decode($token, $cert, ['RS256']);
-            } else {
-                throw new \InvalidArgumentException('cannot parse public key');
-            }
-        } catch (\DomainException $exception) {
-            $this->getLogger()->error('Invalid signature', [$exception]);
-            return ValidationResults::InvalidSignature;
-        } catch (\InvalidArgumentException $exception) {
-            $this->getLogger()->error('Invalid public key', [$exception]);
-            return ValidationResults::InvalidPublicKey;
-        } catch (\UnexpectedValueException $exception) {
-            // Expired, before valid, invalid json, etc
-            $this->getLogger()->debug('Invalid token', [$exception]);
-            return ValidationResults::InvalidToken;
+            $decodedToken = $this->decodeToken($token, $publicCert);
+        } catch (InvalidValidationException $e) {
+            return $e->getCode();
         }
 
         if ($scopes === null) {
             return ValidationResults::Success;
-        } elseif (isset($decoded['scopeCount'])) {
+        } elseif (isset($decodedToken['scopeCount'])) {
             // user scopes not included within
             // the JWT as there are too many
-            throw new ScopesNotDefinedException();
+            throw new ScopesNotDefinedException('too many scopes');
         }
 
-        $isSu = in_array('su', $decoded['scopes'], true);
-        $hasScope = count(array_intersect($scopes, $decoded['scopes'])) > 0;
+        $isSu = in_array('su', $decodedToken['scopes'], true);
+        $hasScope = count(array_intersect($scopes, $decodedToken['scopes'])) > 0;
 
         return $isSu || $hasScope
             ? ValidationResults::Success
             : ValidationResults::Unauthorised;
     }
 
+    /**
+     * Validate and decode a JWT token
+     *
+     * @param string $token a token to validate explicitly, if you do not
+     *      specify one the method tries to find one
+     * @param string $rawPublicCert public key to validate the token
+     * @return array decoded token
+     * @throws InvalidValidationException could not validate token
+     */
+    protected function decodeToken($token, $rawPublicCert)
+    {
+        try {
+            // JWT::decode calls openssl_verify which will cause a fatal error
+            // if the certificate is invalid. Calling openssl_pkey_get_public
+            // first ensures that the certificate is valid before progressing.
+            if ($pubCert = openssl_pkey_get_public($rawPublicCert)) {
+                return (array) JWT::decode($token, $pubCert, ['RS256']);
+            }
+
+            throw new \InvalidArgumentException('cannot parse public key');
```

Comment:
You could `throw new InvalidPublicKeyException('invalid key');` here, unless there's a good reason to do otherwise.  Bit more direct.  There don't seem to be other cases of `InvalidArgumentException` listed in `JWT::decode`:

```php

     * @throws DomainException              Algorithm was not provided
     * @throws UnexpectedValueException     Provided JWT was invalid
     * @throws SignatureInvalidException    Provided JWT was invalid because the signature verification failed
     * @throws BeforeValidException         Provided JWT is trying to be used before it's eligible as defined by 'nbf'
     * @throws BeforeValidException         Provided JWT is trying to be used before it's been created as defined by 'iat'
     * @throws ExpiredException             Provided JWT has since expired, as defined by the 'exp' claim
     *
```